lib/ofp-util.c

```diff
@@ -1899,52 +1900,89 @@ ofputil_decode_dump_map_request(struct ol_dump_map_request *msg,
 
     return error;
 }
 
 struct ofpbuf *
 ofputil_encode_dump_map_request(enum ofp_version ofp_version,
-                          const ovs_be16 prog, const ovs_be16 map_id)
+                                const ovs_be16 prog,
+                                const ovs_be16 nb_maps,
+                                const ovs_be16 *map_ids)
 {
     struct ofpbuf *request;
     struct ol_dump_map_request *msg;
-    request = ofpraw_alloc(OFPRAW_NXT_DUMP_MAP_REQUEST, ofp_version, 0);
+    size_t map_ids_size = (size_t)(sizeof(*map_ids) * nb_maps);
+
+    request = ofpraw_alloc(OFPRAW_NXT_DUMP_MAP_REQUEST, ofp_version, map_ids_size);
     ofpbuf_put_zeros(request, sizeof *msg);
+
     msg = request->msg;
     msg->filter_prog = htons(prog);
-    msg->map_id = htons(map_id);
+    msg->nb_maps = htons(nb_maps);
+
+    ofpbuf_put(request, map_ids, map_ids_size);
+
+    ofpmsg_update_length(request);
 
     return request;
 }
 
 struct ofpbuf *
 ofputil_encode_dump_map_reply(struct ol_dump_map_request *msg,
                               const struct ofp_header *oh,
-                              const struct ubpf_map *map,
-                              void *data,
-                              unsigned int nb_elems)
+                              const struct ubpf_map **map,
+                              const ovs_be16 *map_ids,
+                              void **data,
+                              unsigned int *nb_elems)
 {
     struct ofpbuf *output_buffer;
     struct ol_dump_map_reply *dump_map_reply;
 
-    size_t data_size = (size_t) (nb_elems * (map->key_size + map->value_size));
+    size_t all_maps_size = 0;
+    for(int i = 0; i < msg->nb_maps; i++) {
+        all_maps_size += (size_t) sizeof(struct ol_dump_map);
+        all_maps_size += (size_t) (nb_elems[i] * (map[i]->key_size + map[i]->value_size));
+    }
 
     output_buffer = ofpraw_alloc_reply(OFPRAW_NXT_DUMP_MAP_REPLY, oh,
-            data_size);
+                                       all_maps_size);
 
     ofpbuf_put_zeros(output_buffer,
             sizeof(struct ol_dump_map_reply));
 
     dump_map_reply = output_buffer->msg;
     dump_map_reply->filter_prog = msg->filter_prog;
-    dump_map_reply->map_id = msg->map_id;
-    dump_map_reply->key_size = map->key_size;
-    dump_map_reply->value_size = map->value_size;
-    dump_map_reply->nb_elems = nb_elems;
+    dump_map_reply->nb_maps = msg->nb_maps;
+
+    for(int i = 0; i < msg->nb_maps; i++) {
+        struct ol_dump_map *dump_map = data[i];
+
+        dump_map->map_id = map_ids[i];
+        dump_map->key_size = map[i]->key_size;
+        dump_map->value_size = map[i]->value_size;
+        dump_map->nb_elems = nb_elems[i];
 
-    ofpbuf_put(output_buffer, data, data_size);
+        ofpbuf_put(output_buffer, dump_map, sizeof(*dump_map));
+
+        size_t map_data_size = (size_t) (nb_elems[i] * (map[i]->key_size + map[i]->value_size));
+        ofpbuf_put(output_buffer, data[i] + sizeof(*dump_map), map_data_size);
```

Comment:
If we ask for a large number of maps, `ofpbuf_put` is likely to do a lot of reallocations here. Maybe best to use [`ptr = ofpbuf_put_uninit()`](https://github.com/Orange-OpenSource/oko/blob/11208312300f78a015a1f3d13762d01b38ca5507/lib/ofpbuf.c#L358-L369) once instead then copy data directly to `ptr`?

Reply:
I allocate the tail in `ofpraw_alloc_reply`. There is second loop which counts needed amount of bytes. For sure two `ofpbuf_put` are unneccessary so i have connected them. We do a lot of checks but without reallocations  in my opinion. If it is still too big overhead I can change it.

Reply:
Ah, I see. I missed that. LGTM then!